### ISSUE TITLE
New-JiraSession "You cannot call a method on a null-valued expression"

### ISSUE THREAD

**Issue by rhys-intelliflo** [AtlassianPS/JiraPS#84]
When running New-JIRASession I found it was returning nothing.

```
PS C:\windows\system32> New-JiraSession -Credential (Get-Credential)

cmdlet Get-Credential at command pipeline position 1
Supply values for the following parameters:
WARNING: JIRA returned HTTP error  - 
You cannot call a method on a null-valued expression.
At C:\Program Files\WindowsPowerShell\Modules\PSJira\1.2.5.160\Public\New-JiraSession.ps1:92 char:13
+             $readStream = New-Object -TypeName System.IO.StreamReader ...
+             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (:) [], RuntimeException
    + FullyQualifiedErrorId : InvokeMethodOnNull
 

```

**Comment by rhys-intelliflo** [comment]
By inserting the following to enforce TLS1.2 resolved my issue.

`[Net.ServicePointManager]::SecurityProtocol = [Net.SecurityProtocolType]::Tls12`

**Comment by Clijsters** [comment]
You should definitely check `$webResponse` for not being null before using it. If the internet connection is gone or the name is unresolvable, there is no response. Neither in `$webResponse` directly, nor in `$err.Exception.Response`.

**Comment by lipkau** [comment]
Is this still a problem?

**Comment by itokamak** [comment]
Not sure what's needed to circumvent this? Hitting the same error.

**Comment by rhys-intelliflo** [comment]
Try inserting "[Net.ServicePointManager]::SecurityProtocol = [Net.SecurityProtocolType]::Tls12" before initiating the JIRA session. This worked for me

**Comment by itokamak** [comment]
I'm using it to transition an issue in a script. Adding the above didn't do anything. 
I have it set as: 
[Net.ServicePointManager]::SecurityProtocol = [Net.SecurityProtocolType]::Tls12
#Load Jira Credentials 
$Jira_Username = "psjira"
$Jira_Password = cat C:\scripts\path\to\text\jira.txt | convertto-securestring 
$Jira_Cred = new-object -typename System.Management.Automation.PSCredential -argumentlist $Jira_Username, $Jira_Password 
Set-JiraConfigServer -Server 'https://jira.contoso.com'
New-JiraSession -Credential $Jira_Cred

WARNING: JIRA returned HTTP error  - 
You cannot call a method on a null-valued expression.

**Comment by rhys-intelliflo** [comment]
ah ok .. more debugging will be required ;)

**Comment by kristianlee** [comment]
I'm having the same issue - didn't get it on Powershell v5 in development, but when I run it on a production Powershell v3 host it happens (the version of Windows is also different, so perhaps a red herring). 

Any leads on a workaround / fix @lipkau (the TLS one above doesn't fix it for me either)?

**Comment by lipkau** [comment]
I am sorry, but there is no ETA on this.

I am not able to reproduce the error.
@rhys-intelliflo , @data-source : what powershell version have you been using? can you confirm that this does not happen on PowerShell v5?

Maybe a some debugging on the `New-JiraSession` could help.... ie: what is in `$webResponse` after debug message _"[New-JiraSession] Converting result to JiraSession object"_

Also: (@kristianlee) does this also occur on PoShv3 when not using session but `-Credential`?

**Comment by MichaelChristopherson** [comment]
Ran into this same issue myself.  First time poster so ill post what I have not sure on etiquette.  

I am able to reproduce the issue on command and also know what caused the issue.  I started running into this same issue today after we disabled the TLS 1.0 protocols on the server last night.  The above poster who stated to use:

[Net.ServicePointManager]::SecurityProtocol = [Net.SecurityProtocolType]::Tls12

Directly above the "new-jirasession" command did indeed work.

# **_Here is what the debugging before fixing:_**

New-JiraSession -Credential $Credential -Verbose -Debug
DEBUG: [New-JiraSession] Reading Jira server from config file
Confirm
Continue with this operation?
[Y] Yes [A] Yes to All [H] Halt Command [S] Suspend [?] Help (default is "Yes"): y
DEBUG: [Get-JiraConfigServer] Found Server element. Outputting.
Confirm
Continue with this operation?
[Y] Yes [A] Yes to All [H] Halt Command [S] Suspend [?] Help (default is "Yes"): y
DEBUG: [New-JiraSession] Preparing for blastoff!
Confirm
Continue with this operation?
[Y] Yes [A] Yes to All [H] Halt Command [S] Suspend [?] Help (default is "Yes"): y
VERBOSE: GET https://jira."MyDomainHere".com/rest/api/2/mypermissions with 0-byte payload
DEBUG: [New-JiraSession] Encountered an exception from the Jira server: The request was aborted: Could not create SSL/TLS secure channel.
Confirm
Continue with this operation?
[Y] Yes [A] Yes to All [H] Halt Command [S] Suspend [?] Help (default is "Yes"): y
Cannot index into a null array.
At C:\Program Files\WindowsPowerShell\Modules\JIRAPS\2.3.160\Public\New-JiraSession.ps1:88 char:17
                 $webResponse.Headers[$headerRequiresCaptcha] -and
                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     CategoryInfo          : InvalidOperation: (:) [], RuntimeException
     FullyQualifiedErrorId : NullArray
 
 
 
 
 
 
 
 
 
 
 
#  **_After:_**


[Net.ServicePointManager]::SecurityProtocol = [Net.SecurityProtocolType]::Tls12
New-JiraSession -Credential $Credential -Verbose -Debug
DEBUG: [New-JiraSession] Reading Jira server from config file
Confirm
Continue with this operation?
[Y] Yes [A] Yes to All [H] Halt Command [S] Suspend [?] Help (default is "Yes"): y
DEBUG: [Get-JiraConfigServer] Found Server element. Outputting.
Confirm
Continue with this operation?
[Y] Yes [A] Yes to All [H] Halt Command [S] Suspend [?] Help (default is "Yes"): y
DEBUG: [New-JiraSession] Preparing for blastoff!
Confirm
Continue with this operation?
[Y] Yes [A] Yes to All [H] Halt Command [S] Suspend [?] Help (default is "Yes"): y
VERBOSE: GET https://jira."MyDomainHere".com/rest/api/2/mypermissions with 0-byte payload
VERBOSE: received -1-byte response of content type application/json;charset=UTF-8
DEBUG: [New-JiraSession] Converting result to JiraSession object
Confirm
Continue with this operation?
[Y] Yes [A] Yes to All [H] Halt Command [S] Suspend [?] Help (default is "Yes"): y
DEBUG:
public class JsonObjectTypeResolver : System.Web.Script.Serialization.JavaScriptTypeResolver
{
        public override System.Type ResolveType(string id)
        {
                return typeof (System.Collections.Generic.Dictionary<string, object>);
        }
 
        public override string ResolveTypeId(System.Type type)
        {
                return string.Empty;
        }
}
Confirm
Continue with this operation?
[Y] Yes [A] Yes to All [H] Halt Command [S] Suspend [?] Help (default is "Yes"): y
DEBUG: [New-JiraSession] Saving session in module's PrivateData
Confirm
Continue with this operation?
[Y] Yes [A] Yes to All [H] Halt Command [S] Suspend [?] Help (default is "Yes"): y
DEBUG: [New-JiraSession] Adding session result to existing module PrivateData
Confirm
Continue with this operation?
[Y] Yes [A] Yes to All [H] Halt Command [S] Suspend [?] Help (default is "Yes"): y
DEBUG: [New-JiraSession] Outputting result
Confirm
Continue with this operation?
[Y] Yes [A] Yes to All [H] Halt Command [S] Suspend [?] Help (default is "Yes"): y
 
Username        WebSession                                      JSessionID LoginInfo
--------        ----------                                      ---------- ---------
"My Username"   Microsoft.PowerShell.Commands.WebRequestSession


Issue was reproduced on the following powershell versions:
5.1.15063.483
5.1.14409.1005

I did not attempt on any other versions.

Let me know if I can help gather any other useful data.

**Comment by rhys-intelliflo** [comment]
HI @lipkau I've been using PSv5  and now 5.1 and issue remains.
I get it by 
`New-JiraSession -Credential (Get-Credential)`
Then I get 

> Cannot index into a null array.
> At C:\Program Files\WindowsPowerShell\Modules\JIRAPS\2.3.160\Public\New-JiraSession.ps1:88 char:17
> +                 $webResponse.Headers[$headerRequiresCaptcha] -and
> +                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
>     + CategoryInfo          : InvalidOperation: (:) [], RuntimeException
>     + FullyQualifiedErrorId : NullArray

**Comment by lipkau** [comment]
Please update your version

**Comment by rhys-intelliflo** [comment]
Hi @lipkau 
Have done. Now runnng 2.4.0 of JiraPS and same thing. I have tested this on a freshly built machine and issue remains.

**Comment by lipkau** [comment]
In that case, please enter dubugging and show me what `convertto-json
$webResponse` looks like

**Comment by Dejulia489** [comment]
This exception is caught in my latest pull request.

**Comment by lipkau** [comment]
Fixed with version 2.5

If you still encounter this problem, please create a new issue and link to this one
